nutils/pointsseq.py

```diff
@@ -0,0 +1,648 @@
+
+  @property
+  def npoints(self):
+    return self._parent.npoints * self._count
+
+  def __len__(self):
+    return len(self._parent)*self._count
+
+  def __iter__(self):
+    for i in range(self._count):
+      yield from self._parent
+
+  def get(self, index):
+    return self._parent.get(numeric.normdim(len(self), index) % len(self._parent))
+
+  def repeat(self, count):
+    if count == 0:
+      return Empty(self.ndims)
+    else:
+      return Repeat(self._parent, self._count * count)
+
+  def _mk_indices(self, parent):
+    npoints = self._parent.npoints
+    ind = parent[None] + numpy.arange(0, self._count*npoints, npoints)[:,None,None]
+    ind = ind.reshape(self._count*parent.shape[0], parent.shape[1])
+    return types.frozenarray(ind, copy=False)
+
+  @property
+  def tri(self):
+    return self._mk_indices(self._parent.tri)
+
+  @property
+  def hull(self):
+    return self._mk_indices(self._parent.hull)
+
+class Product(PointsSequence):
+  '''The product of two sequences.
+
+  Parameters
+  ----------
+  sequence1 : :class:`PointsSequence`
+  sequence2 : :class:`PointsSequence`
+  '''
+
+  __slots__ = '_sequence1', '_sequence2'
+
+  @types.apply_annotations
+  def __init__(self, sequence1: strictpointsseq, sequence2: strictpointsseq):
+    if sequence1.isuniform and sequence2.isuniform:
+      warnings.warn('both sequences are uniform; use `Uniform(sequence1.get(0) * sequence2.get(0), len(references1) * len(references2))` instead', warnings.NutilsInefficiencyWarning)
+    self._sequence1 = sequence1
+    self._sequence2 = sequence2
+    super().__init__(sequence1.ndims + sequence2.ndims)
+
+  @property
+  def npoints(self):
+    return self._sequence1.npoints * self._sequence2.npoints
+
+  def __len__(self):
+    return len(self._sequence1) * len(self._sequence2)
+
+  def __iter__(self):
+    return (item1.product(item2) for item1 in self._sequence1 for item2 in self._sequence2)
+
+  def get(self, index):
+    index1, index2 = divmod(numeric.normdim(len(self), index), len(self._sequence2))
+    return self._sequence1.get(index1).product(self._sequence2.get(index2))
+
+  def product(self, other):
+    return self._sequence1.product(self._sequence2.product(other))
+
+class Chain(PointsSequence):
+  '''A chain of two sequences.
+
+  Parameters
+  ----------
+  sequence1 : :class:`PointsSequence`
+  sequence2 : :class:`PointsSequence`
+  '''
+
+  __slots__ = '_sequence1', '_sequence2'
+  __cache__ = 'tri', 'hull'
+
+  @types.apply_annotations
+  def __init__(self, sequence1: strictpointsseq, sequence2: strictpointsseq):
+    if sequence1.ndims != sequence2.ndims:
+      raise ValueError('cannot chain sequences with different ndims')
+    elif sequence1.isuniform and sequence2.isuniform and sequence1.get(0) == sequence2.get(0):
+      warnings.warn('both sequences are uniform sequences of the same item, use `Uniform(sequence1.get(0), len(sequence1) + len(sequence2))` instead', warnings.NutilsInefficiencyWarning)
```

Comment:
we can also warn for the situation sequence1 == sequence2: should be a Repeat.

Reply:
done